### PR TITLE
Add alexa event source

### LINKED ISSUE #2875: Add support for Alexa as an event source
# This is a Feature Proposal

## Description

Update @rajington great PR (https://github.com/serverless/serverless/pull/1525) / plugin (https://github.com/rajington/serverless-alexa-plugin) so that Alexa as an event source is supported.

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #2875 

Adds the `alexaSkill` event definition.

Shoutouts to @rajington who implemented the first proposal  🙌 ⚗ 

## How did you implement it:

This adds a permission template to the Lambda function based on https://github.com/serverless/serverless/pull/1525, https://github.com/rajington/serverless-alexa-plugin and https://developer.amazon.com/public/community/post/Tx27NAUCY0KQ34D/Rapidly-Create-Your-Alexa-Skill-Backend-with-AWS-CloudFormation

## How can we verify it:

```yml
service: alexa-test-service

provider:
  name: aws
  runtime: nodejs4.3
  cfLogs: true

functions:
  hello:
    handler: handler.hello
    events:
      - alexaSkill
```

Run `serverless deploy --noDeploy`. Alternatively run the test suite...

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

**Comment by pmuens** [comment]
**Open questions:**

- [ ] Should we also lock down the permissions here? If so, how?
- [ ] Should we switch to an object representation so that this event can be extended more easily?
- [ ] Should this event be named `alexa` or something else (such as `ask`)?

Here's a proposal for the object representation:

```yml
service: alexa-test-service

provider:
  name: aws
  runtime: nodejs4.3
  cfLogs: true

functions:
  hello:
    handler: handler.hello
    events:
      - alexa:
          enabled: true
```

Right now it only has the `enabled` property which will simply add the permission. This way we can extend this event way easier later on. Not sure if a simple representation (like `alexa: true`) is necessary if we go this way?

What do you think @mthenw @eahefnawy @nikgraf ?

**Comment by nikgraf** [comment]
- Should we also lock down the permissions here? If so, how?

I can't tell. How do the permissions work here?

- Should we switch to an object representation so that this event can be extended more easily?

Yes, Amazon might add more options and we should stay flexible.

- Should this event be named alexa or something else (such as ask)?

In the SAM they used `AlexaSkill` as event type. I recommend we do the same as Alexa become more powerful and new event types might be introduced in the future. 

- Right now each function can only have one alexa event. Are there any use cases where one function might need more than one alexa event?

Why can there only be one?

**Comment by pmuens** [comment]
Thanks for the feedback @nikgraf I agree with everything. Will update it soon!

> Why can there only be one?

Because we have the `alexa: true` event definition. Adding another `alexa: true` is not possible becaus they both have the same key. We need something like the skill set name to identify those (although CloudFormation does not need it as it only adds a "global" permission template)...

**Comment by nikgraf** [comment]
@pmuens is this ready for review?

**Comment by pmuens** [comment]
@nikgraf woops. Forgot to update the message. Thanks!

Yes, this can be reviewed!
Some of the questions we already discussed needs to be addressed. Waiting for @eahefnawy and @mthenw feedback on this.

**Comment by mthenw** [comment]
> Right now it only has the enabled property which will simply add the permission. This way we can extend this event way easier later on. Not sure if a simple representation (like alexa: true) is necessary if we go this way?

IMHO current implementation is fine. 

> Should we also lock down the permissions here? If so, how?

Can we lock it down any how?

**Comment by pmuens** [comment]
> Can we lock it down any how?

Not sure if we can do this. We need to set the `SourceArn` property in the permission (like it's done here for the S3 event: https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/events/s3/index.js#L168-L172). However we don't have a source which is a CloudFormation resource. So I'm not quite sure if it's even possible.

Maybe @eahefnawy can shed some light into this one as he's locked down the permissions for the other events as well.

---

The other open question is if one Lambda function can have multiple `alexaSkill` events.\

The problem we have is that we cannot identify the alexa skill event correctly since we don't have a name of the skill, or something else which might be unique.

**Comment by nikgraf** [comment]
Just discussed with @pmuens. Narrowing down the permissions is not possible, because currently AlexaSkills don't expose any ARNs. You just can allow to expose the Lambda to the Alex portal.

We will drop `enabled: true` since it's an arbitrary property we would defined, but we don't have it in any other event nor it is available in AWS.

File: lib/plugins/aws/deploy/compile/events/alexaSkill/index.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      'deploy:compileEvents': this.compileAlexaSkillEvents.bind(this),
+    };
+  }
+
+  compileAlexaSkillEvents() {
+    this.serverless.service.getAllFunctions().forEach((functionName) => {
+      const functionObj = this.serverless.service.getFunction(functionName);
+
+      if (functionObj.events) {
+        console.log(functionObj.events);
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/alexaSkill/index.js]: This should be removed

**Comment by pmuens** [comment]
Great @nikgraf! Thanks for updating this PR!

Looks good from my side! 👍

**Pull request merged**